cstar/base/input_dataset.py

```diff
@@ -1,27 +1,103 @@
 import os
 import pooch
 import hashlib
+import pathlib
+from abc import ABC
 import datetime as dt
 import dateutil.parser
+from urllib.parse import urlparse
 from typing import Optional, TYPE_CHECKING
 
-from cstar.base.utils import _get_source_type
-
-
 if TYPE_CHECKING:
     from cstar.base import BaseModel
 
 
-class InputDataset:
+class DataSource:
+    """
+    Describes the source of an InputDataset
+
+    Attributes:
+    -----------
+    location: str
+       The location of the data, e.g. a URL or local path
+
+    Properties:
+    -----------
+    location_type: str (read-only)
+       "url" or "path"
+    source_type: str (read only)
+       Typically describes file type (e.g. "netcdf") but can also be "repository"
+    basename: str (read-only)
+       The basename of self.location, typically the file name
+    """
+
+    def __init__(self, location: str):
+        """
+        Initialize a DataSource from a location string
+
+        Parameters:
+        -----------
+        location: str
+           The location of the data, e.g. a URL or local path
+
+        Returns:
+        --------
+        DataSource
+            An initialized DataSource
+        """
+        self.location = location
+
+    @property
+    def location_type(self) -> str:
+        """Get the location type (e.g. "path" or "url") from the "location" attribute"""
+        urlparsed_location = urlparse(self.location)
+        if all([urlparsed_location.scheme, urlparsed_location.netloc]):
+            return "url"
+        elif pathlib.Path(self.location).exists():
+            return "path"
+        else:
+            raise ValueError(
+                f"{self.location} is not a recognised URL or local path pointing to an existing file"
+            )
+
+    @property
+    def source_type(self) -> str:
+        """Get the source type (e.g. "netcdf" from the "location" attribute"""
+        if self.location.lower().endswith((".yaml", ".yml")):
+            return "yaml"
+        elif self.location.lower().endswith(".nc"):
+            return "netcdf"
+        elif self.location.lower().endswith(".git"):
+            return "repository"
+        else:
+            raise ValueError(
+                f"{os.path.splitext(self.location)[-1]} is not a supported file type"
+            )
+
+    @property
+    def basename(self) -> str:
+        """Get the basename (typically a file name) from the location attribute"""
+        return os.path.basename(self.location)
+
```

Comment:
This suggestion isn't really important, but it demonstrates the difference in how `__repr__` is recommended to be used vs how `__str__` is recommended to be used. Ideally `call(repr(obj)) == obj`, whereas `__str__` can be more verbose.

```suggestion
def __repr__(self) -> str:
    return f"DataSource(location={self.location})"

```